src/dag.rs

```diff
@@ -30,43 +174,414 @@ impl<Types: RepoTypes> IpldDag<Types> {
         Ok(cid)
     }
 
-    pub async fn get(&self, path: IpfsPath) -> Result<Ipld, Error> {
+    /// Resolves a Cid rooted path to a document "node."
+    ///
+    /// Returns the resolved node as Ipld.
+    pub async fn get(&self, path: IpfsPath) -> Result<Ipld, ResolveError> {
+        // FIXME: do ipns resolve first
         let cid = match path.root().cid() {
             Some(cid) => cid,
-            None => return Err(anyhow::anyhow!("expected cid")),
+            None => panic!("not implemented: Ipns resolution, expected Cid based path"),
+        };
+
+        let mut iter = path.iter().peekable();
+
+        let (node, _) = match self.resolve0(cid, &mut iter, true).await {
+            Ok(t) => t,
+            Err(e) => {
+                drop(iter);
+                return Err(e.with_path(path));
+            }
         };
-        let mut ipld = decode_ipld(&cid, self.ipfs.repo.get_block(&cid).await?.data())?;
-        for sub_path in path.iter() {
-            ipld = try_resolve(ipld, sub_path)?;
-            ipld = match ipld {
-                Ipld::Link(cid) => decode_ipld(&cid, self.ipfs.repo.get_block(&cid).await?.data())?,
-                ipld => ipld,
+
+        Ipld::try_from(node)
+    }
+
+    /// Resolves a Cid rooted path to a document "node."
+    ///
+    /// The return value has two kinds of meanings depending on whether links should be followed or
+    /// not: When following links the second returned value will be the path inside last document.
+    /// When not following links the second returned value will be the unmatched or "remaining"
+    /// path.
+    ///
+    /// Regardless of the `follow_links` option, HAMT sharded directories will be resolved through
+    /// as a "single step" in the given IpfsPath.
+    ///
+    /// Returns the node and remaining path or the path inside the last document.
+    pub async fn resolve(
+        &self,
+        path: IpfsPath,
+        follow_links: bool,
+    ) -> Result<(ResolvedNode, SlashedPath), ResolveError> {
+        // FIXME: do ipns resolve first
+        let cid = match path.root().cid() {
+            Some(cid) => cid,
+            None => panic!("not implemented: Ipns resolution, expected Cid based path"),
+        };
```

Comment:
nit: this one (and the one in `get` as well) could just be `let cid = path.root().cid().expect("not implemented: Ipns resolution, expected Cid based path");`

Reply:
yeah this could be written as such. I'd really prefer these methods to be about `CidPath`s though and `Ipns::resolve(IpfsPath) -> Result<CidPath, _>`. I should record this down into fixme though.